BROOKLYN-356: fix race in transformer by adding getImmediately

aledsage opened the pull request.

geomacy commented on core/src/main/java/org/apache/brooklyn/util/core/task/ValueResolver.java:
```diff
@@ -300,6 +313,18 @@ public T get() {
         
         try {
+            if (immediately && v instanceof ImmediateSupplier) {
+                final ImmediateSupplier<?> supplier = (ImmediateSupplier<?>) v;
```
Does the comment at line 383 need updated now that this is added?

geomacy commented on core/src/test/java/org/apache/brooklyn/core/entity/EntitySubscriptionTest.java:
```diff
@@ -280,4 +279,55 @@ public void testSubscriptionForInitialValueWhenNotValid() {
+        entity.subscriptions().subscribeToChildren(observedEntity, TestEntity.SEQUENCE, listener);
+        observedChildEntity.sensors().set(TestEntity.SEQUENCE, 123);
+        assertListenerCalledOnceWithContextEntityEventually(listener, entity);
+        listener.clearEvents();
```
should this be in a `tearDown`?

geomacy commented on core/src/test/java/org/apache/brooklyn/util/core/task/ValueResolverTest.java:
```diff
@@ -41,6 +49,125 @@ public void setUp() throws Exception {
+        Assert.assertEquals(result.get(), "foo");
+    }
+
+    public void testNoExecutionContextOnCompleted() {
```
Don't follow what this is testing - how is it testing something about 'no execution context'?

aledsage commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java:
```diff
@@ -353,6 +386,33 @@ public DslSensorSupplier(DslComponent component, Object sensorIndicator) {
+                    result = targetEntity.getEntityType().getSensor((String)si);
+                }
+                if (result!=null) return Maybe.<Sensor<?>>of(result);
+                return Maybe.<Sensor<?>>of(Sensors.newSensor(Object.class, (String)si));
```
Not sure I follow. In `attributeWhenReady`, if we're given a sensor name then we will instantiate a new sensor object if it's not defined on the entity type; and then we look up the sensor's value (returning Maybe.absent if it's null). Here, we just want it to return the sensor object (rather than a sensor value) so we do the first half: instantiate a new sensor object if it's not defined on the entity type.

This is consistent with the blocking `get()` method. We return `Maybe.absent()` if the `get()` method would have blocked, and otherwise we return the same value as `get()` would have returned.

aledsage commented on camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/dsl/DslTest.java:
```diff
@@ -118,39 +160,64 @@ public void runConcurrentWorker(Supplier<Runnable> taskSupplier) {
-            for (int i = 0; i < RESOLVER_ITERATIONS; i++) {
+            TestEntity entity = parent.addChild(childSpec);
+            for (int i = 0; i < resolverIterations; i++) {
                 preResolve(entity);
```
I'll rename this from "eventually" to "satisfiedAsynchronously".

What "eventually" meant was that the condition for getting the value immediately is not yet satisfied; it will become satisfied asynchronously. For example, when testing `attributeWhenReady` then another thread sets the sensor's value after we have called `execDslEventually` - we expect that method to block.

I'll also change the assertions so that we confirm we get a Maybe or the real result when calling getImmediately in this situation.

aledsage commented:
Thanks @geomacy for the great comments - I've addressed those in an additional commit, and rebased against master.

Once jenkins confirms the tests pass, I'll merge. If you do get a chance to review the changes in the last commit before that, it would be great though.

neykov commented:
@aledsage I'm currently reviewing the PR, please hold on with the merge.

neykov commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java:
```diff
@@ -525,15 +583,20 @@ public DslExternal(String providerName, String key) {
         @Override
+        public final Maybe<Object> getImmediately() {
+            ManagementContextInternal managementContext = DslExternal.managementContext();
+            return Maybe.<Object>of(managementContext.getExternalConfigProviderRegistry().getConfig(providerName, key));
```
This is not reusable since `getConfig` will block indefinitely, resolving the value.

aledsage replied:
@neykov This is an interesting one! The `managementContext.getExternalConfigProviderRegistry().getConfig(...)` is calling a different getConfig whose javadoc simply says:

```
    /**
     * Searches the named {@link ExternalConfigSupplier} for the config value associated with the specified key.
     * Quietly returns <code>null</code> if no config exists for the specified key.
     * Throws {@link IllegalArgumentException} if no {@link ExternalConfigSupplier} exists for the passed name.
     */
```

The existing implementation delegates to `ExternalConfigSupplier.get(key)`.

There is no guidance on that whether the call could be blocking, or should return reasonably promptly.

I'd assume it would call out to things like Vault, and would return the value that currently exists or null. Such calls will not be super-fast, but should not be blocked waiting for other entities either.

I therefore think this is good enough, to just call getConfig without wrapping it in another task for timeout-purposes.

If it's not good enough, then we should really change `DslExternal` back so that it doesn't even try to support `getImmediately()`. Or we could try to change the interface of `ExternalConfigSupplierRegistry` etc.

neykov replied:
Agree - it's fine calling this in `getImmediately`.

neykov commented on core/src/test/java/org/apache/brooklyn/util/core/task/ValueResolverTest.java:
```diff
@@ -69,46 +59,65 @@ public void testTimeoutBig() {
+        Maybe<String> result = Tasks.resolving(t).as(String.class).timeout(Duration.ZERO).getMaybe();
+        
+        Assert.assertTrue(result.isAbsent(), "result="+result);
+        Exception exception = ((Maybe.Absent<?>)result).getException();
```
Can use `Maybe.getException(result)`.

neykov commented:
There's pretty big overlap between the `getImmediate` & `newTask` implementations - would be nice to reuse as much as possible, but I'm fine merging those as is.

geomacy commented on core/src/test/java/org/apache/brooklyn/util/core/task/ValueResolverTest.java:
```diff
@@ -69,46 +59,65 @@ public void testTimeoutBig() {
 
-    public void testNoExecutionContextOnCompleted() {
+    public void testCompletedTaskReturnsResultImmediately() {
+        // Call ValueResolver.getMaybe() from this thread, which has no execution context.
```
I still struggled with understanding what was meant here until I realised that this line 63 refers to the `Tasks.resolving(t)` being executed in this thread (I'd been put off by concentrating on the line `app.getExecutionContext()...`).  It would maybe make this clearer if these lines of comment were moved to just above the `Tasks.resolving(t)` on 68 below.

geomacy commented on core/src/test/java/org/apache/brooklyn/util/core/task/ValueResolverTest.java:
```diff
@@ -124,4 +133,141 @@ public void testDefaultBeforeDelayAndError() {
+        MyImmediateAndDeferredSupplier supplier = new MyImmediateAndDeferredSupplier();
+        CallInfo callInfo = Tasks.resolving(supplier).as(CallInfo.class).context(app).timeout(Asserts.DEFAULT_LONG_TIMEOUT).get();
+        assertNotNull(callInfo.task);
+        assertNotContainsCallingMethod(callInfo.stackTrace, "testGetImmediately");
```
The test is called `testGetImmediateXXXX` not `testGetImmediatelyXXXX`.  In any case it might be worth supplying the full method name here to avoid possible confusion with line 140. Same with line 170 below.

geomacy commented:
Couple more comments but otherwise looks good along with Svet's comments.  👍

aledsage commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java:
```diff
@@ -303,6 +330,12 @@ public DslConfigSupplier(DslComponent component, String keyName) {
         @Override
+        public final Maybe<Object> getImmediately() {
+            Entity targetEntity = component.get();
+            return Maybe.of(targetEntity.getConfig(ConfigKeys.newConfigKey(Object.class, keyName)));
```
Great spot! I'm writing a bunch more tests for this.

aledsage commented:
@neykov @geomacy latest comments addresses in the last two commits. Can you take another look please?

neykov commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java:
```diff
@@ -340,9 +360,9 @@ public DslConfigSupplier(DslComponent component, String keyName) {
+            EntityInternal targetEntity = (EntityInternal) targetEntityMaybe.get();
             
-            return Maybe.of(targetEntity.getConfig(ConfigKeys.newConfigKey(Object.class, keyName)));
+            return targetEntity.config().getNonBlocking(ConfigKeys.newConfigKey(Object.class, keyName));
```
Nice, forgot we have this method. Now take a guess at how it's implemented :)
The [simple](https://github.com/apache/brooklyn-server/blob/master/core/src/main/java/org/apache/brooklyn/core/objs/AbstractConfigurationSupportInternal.java#L133) case looks easy enough to update. The [structure key](https://github.com/apache/brooklyn-server/blob/master/core/src/main/java/org/apache/brooklyn/core/objs/AbstractConfigurationSupportInternal.java#L91) case looks like a rabbit hole though (with an all too familiar comment).
I'm fine with leaving the structured key case as is for this PR, may be log a warning.

aledsage replied:
Yes, I'd updated `AbstractConfigurationSupportInternal` line 142 to use `.immediately(true)`, so I hope that covers the case of `getNonBlockingResolvingSimple`. I didn't want to touch the "structure key" code. We should probably change that at some point to go through ValueResolver, even if that still ends up scheduling it in a task.

I therefore think this comment has been addressed. Am I missing something?

neykov replied:
Somehow missed that. All good then.

neykov commented:
LGTM, merging.

geomacy commented on rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/AbstractBrooklynRestResource.java:
```diff
@@ -143,11 +145,12 @@ public Object resolve() {
             return Tasks.resolving(value)
                     .as(Object.class)
                     .defaultValue(UNRESOLVED)
                     .timeout(timeout)
```
What happens with a timeout _and_ an immediately? Should this be `.timeout( immediately ? 0 : timeout)`? (with adjustments for Boolean)

geomacy commented:
One minor comment in AbstractBrooklynRestResource.java but overall LGTM 👍 . I had a look at the code @neykov referred to and I suppose there is more to do as noted in the comment in `getNonBlockingResolvingStructuredKey` but I agree that that should not be part of this change.
